File: Euler-Go/src/problem_23.go
package main

import "fmt"

func main() {
	fmt.Println("Greg is cool!")
	fmt.Println(sum_all_integers())
	//fmt.Println(sum(abundant_numbers()))	
	//fmt.Println(proper_divisors(28))
	//fmt.Println(sum(proper_divisors(28)))
	//fmt.Println(sum_all_integers() - sum(abundant_numbers()))
	fmt.Println(sum(abundant_sums(abundant_numbers())))
}


func abundant_sums(aN []int) []int{
	counter := 0
	zzz := 0
	aS := make([]int, 1000000)
	for i := range aN {
	//for i := 0; i <= len(aN); i ++ {
		for j := range aN {
		//for j := 0; j <= len(aN); j ++ {
			fmt.Println("aN[i] = ", aN[i])
			fmt.Println("aN[j] = ", aN[j])
			zzz = aN[i] + aN[j]
			if zzz < 28123 {
				aS[counter] = zzz
				counter += 1
			}
		}
	}
	return aS[:counter]
}

func abundant_numbers() []int{
	number := 281
	counter := 1
	L := make([]int, number)
	for i := 1; i <= number; i ++{
		if sum(proper_divisors(i)) > i{
			L[counter] = i
			counter += 1
		}
	}
	return L[:counter]
}

func sum(pd []int) (int){
	sum := 0
	for i := range pd {
		sum += pd[i]
	}
	return sum
}

func proper_divisors(number int) (factors []int) {
	L := make([]int, number)
	counter := 0
	for i := 1; i < ((number / 2) + 1); i++ {
		if number%i == 0 {
			L[counter] = i
			counter += 1
		}
	}
	return L[:counter]
}

func sum_all_integers() int {
	sai := 0
	for i := 1; i < 28123; i ++ {
		sai += i
	}
	return sai
}

File: Euler-Java/src/Problem_23.java
import java.util.ArrayList;
import java.util.List;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.Collection;

public class Problem_23 {
	public static void main(String args[]){
		
		long start = System.currentTimeMillis();
		
		List<Integer> aN = new ArrayList<Integer>(Abundant_numbers());
		Collections.sort(aN);
		
		Collection<Integer> abundant_sums = new ArrayList<Integer>(Abundant_sums(aN));
		Collection<Integer> all_integers = new ArrayList<Integer>(All_integers());
		
		all_integers.removeAll(abundant_sums);
		
		System.out.println(sum_Answer(all_integers));
		
		long stop = System.currentTimeMillis();
		System.out.println("Run time = " + (stop - start) + " milliseconds");
	}
	
	public static Integer sum_Answer(Collection<Integer> list) {
	     Integer sum= 0;
	     if(list == null){
	    	 return 0;
	     }
	     for (Integer i:list)
	         sum = sum + i;
	     return sum;
	}
	
	public static Set<Integer> All_integers(){
		Set<Integer> All_integers = new HashSet<Integer>();
		for(int i = 1; i < 28123; i ++){
			All_integers.add(i);
		}
		return All_integers;
	}
	
	
	public static Set<Integer> Abundant_numbers(){
		Set<Integer> Abundant_numbers = new HashSet<Integer>();
		for(int i = 1; i <= 28123; i ++){
			List<Integer> x = proper_divisors(i);
			if(sum(x) > i){
				Abundant_numbers.add(i);
			}
		}
		return Abundant_numbers;
	}
	
	public static Set<Integer> Abundant_sums(List<Integer> aN){
		Set<Integer> Abundant_sums = new HashSet<Integer>();
		for (int i = 0; i < aN.size(); i++) {
			Integer x = aN.get(i);
			for(int j = 0; j < aN.size(); j++){
				int y = aN.get(j);
				int zzz = x + y;
				if(zzz < 28123){
					Abundant_sums.add(zzz);
				}
			}
		}
		return Abundant_sums;
	}

	public static List<Integer> proper_divisors(int number){
		List<Integer> L = new ArrayList<Integer>();
		if(number == 1){
			L = null;
			return L;
		}
		if(number == 2){
			L.add(1);
			return L;
		}
		for(int i = 1; i <= ((number / 2) + 1); i ++){
			if(number % i == 0){
				L.add(i);
			}
		}
		return L;
	}
	
	public static Integer sum(List<Integer> list) {
	     Integer sum= 0;
	     if(list == null){
	    	 return 0;
	     }
	     for (Integer i:list)
	         sum = sum + i;
	     return sum;
	}
}
